File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-csharp/src/SDKWork.Mail.Sdk/mailProviderCatalog.cs
namespace Sdkwork.Mail.Sdk;

using System.Linq;

public sealed record MailProviderCatalogEntry(
    string providerKey,
    string pluginId,
    string driverId,
    bool defaultSelected
);

public static class MailProviderCatalog
{
    public const string DEFAULT_mail_PROVIDER_KEY = "smtp";

    public static readonly IReadOnlyList<MailProviderCatalogEntry> Entries =
    [
        new("smtp", "Mail-smtp", "sdkwork-mail-driver-smtp", true),
        new("imap", "Mail-imap", "sdkwork-mail-driver-imap", false),
    ];

public static MailProviderCatalogEntry? GetMailProviderByProviderKey(string providerKey) =>
        Entries.FirstOrDefault(entry => entry.providerKey == providerKey);

}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-flutter/lib/src/mail_provider_catalog.dart
final class MailProviderCatalogEntry {
  const MailProviderCatalogEntry({
    required this.providerKey,
    required this.pluginId,
    required this.driverId,
    required this.defaultSelected,
  });

  final String providerKey;
  final String pluginId;
  final String driverId;
  final bool defaultSelected;
}

final class MailProviderCatalog {
  static const String DEFAULT_mail_PROVIDER_KEY = "smtp";

  static const List<MailProviderCatalogEntry> entries = <MailProviderCatalogEntry>[
    MailProviderCatalogEntry(
      providerKey: "smtp",
      pluginId: "Mail-smtp",
      driverId: "sdkwork-mail-driver-smtp",
      defaultSelected: true,
    ),
    MailProviderCatalogEntry(
      providerKey: "imap",
      pluginId: "Mail-imap",
      driverId: "sdkwork-mail-driver-imap",
      defaultSelected: false,
    )
  ];

  const MailProviderCatalog._();
}

MailProviderCatalogEntry? getMailProviderByProviderKey(String providerKey) {
  for (final entry in MailProviderCatalog.entries) {
    if (entry.providerKey == providerKey) {
      return entry;
    }
  }

  return null;
}
